src/mlpack/methods/decision_tree/mse_gain.hpp

```diff
@@ -96,54 +96,179 @@ class MSEGain
+    }
+    else
+    {
+      // Stores the sum of squares till the previous index.
+      double prevSumSquares = 0.0;
+
+      for (size_t i = 0; i < minimum - 1; ++i)
+      {
+        const RType x = responses[i];
+
+        // Calculating the initial mean of responses for the left child.
+        ++leftSize;
+        leftMean += x;
+        sumSquares[i] = prevSumSquares + x * x;
+        prevSumSquares += x * x;
+      }
+      if (leftSize > 1e-9)
+        leftMean /= leftSize;
+
+      for(size_t i = minimum - 1; i < responses.n_elem; ++i)
+      {
+        const RType x = responses[i];
+
+        // Calculating the initial mean of responses for the right child.
+        ++rightSize;
+        rightMean += x;
+        sumSquares[i] = prevSumSquares + x * x;
+        prevSumSquares += x * x;
+      }
+      if (rightSize > 1e-9)
+        rightMean /= rightSize;
+
+      totalSumSquares = prevSumSquares;
+    }
+  }
+
+  /**
+   * Updates the statistics for the given index.
    *
-   * @param sumSquares Precomputed sum of square (sum(Xi^2)) of values.
-   * @param mean Precomputed mean (sum(Xi) / n) of values.
-   * @param childSize The total number of samples in that child.
+   * @param responses The set of responses on which statistics are computed.
+   * @param weights The set of weights associated to each response.
+   * @param index The current index.
    */
-  static double Evaluate(const double sumSquares,
-                         const double mean,
-                         const size_t childSize)
+  template<bool UseWeights, typename ResponsesType, typename WeightVecType>
+  void UpdateStatistics(const ResponsesType& responses,
+                        const WeightVecType& weights,
+                        const size_t index)
   {
-    double mse = sumSquares / (double) childSize - mean * mean;
-    return -mse;
+    typedef typename ResponsesType::elem_type RType;
+    typedef typename WeightVecType::elem_type WType;
+
+    if (UseWeights)
+    {
+      const WType w = weights[index];
+      const RType x = responses[index];
+      leftMean = (leftMean * leftSize + w * x) / (leftSize + w);
+      leftSize += w;
+
+      rightMean = (rightMean * rightSize - w * x) / (rightSize - w);
+      rightSize -= w;
+    }
+    else
+    {
+      const RType x = responses[index];
+      leftMean = (leftMean * leftSize + x) / (leftSize + 1);
+      ++leftSize;
+
+      rightMean = (rightMean * rightSize - x) / (rightSize - 1);
+      --rightSize;
+    }
   }
+
+ private:
+  /**
+   * The following data members cache statistics for weighted data when
+   * `UseWeights` is true, else it will calculate unweighted statistics.
+   */
+  // Stores the sum of squares / weighted sum of squares.
+  arma::rowvec sumSquares;
+  // For unweighted data, stores the number of elements in each child.
+  // For weighted data, stores the sum of weights of elements in each
+  // child.
+  double leftSize;
+  double rightSize;
```

Comment:
This is a minor implementation detail, but if you did hold `lastIndex`, then both `leftSize` and `rightSize` can be derived from it and the total length: `leftSize = lastIndex` and `rightSize = responses.n_elem - leftSize`.

Reply:
These variables also store the sum of weights for left and right children too in weighted training. So, if we remove one of these, then we will have to add another data member that stores the total weight. So, we will end up the same thing.

Reply:
You're right, I forgot about the weighted case. :)